mmdet/models/dense_heads/paa_head.py

```diff
@@ -0,0 +1,586 @@
+        if not len(pos_inds):
+            return label, label_weight, bbox_weight, 0
+
+        num_gt = pos_gt_inds.max() + 1
+        num_level = len(anchors)
+        num_anchors_each_level = [item.size(0) for item in anchors]
+        num_anchors_each_level.insert(0, 0)
+        inds_level_interval = np.cumsum(num_anchors_each_level)
+        pos_level_mask = []
+        for i in range(num_level):
+            mask = (pos_inds >= inds_level_interval[i]) & (
+                pos_inds < inds_level_interval[i + 1])
+            pos_level_mask.append(mask)
+        pos_inds_after_paa = []
+        for gt_ind in range(num_gt):
+            pos_inds_gmm = []
+            pos_loss_gmm = []
+            gt_mask = pos_gt_inds == gt_ind
+            for level in range(num_level):
+                level_mask = pos_level_mask[level]
+                level_gt_mask = level_mask & gt_mask
+                value, topk_inds = pos_losses[level_gt_mask].topk(
+                    min(level_gt_mask.sum(), self.topk), largest=False)
+                pos_inds_gmm.append(pos_inds[level_gt_mask][topk_inds])
+                pos_loss_gmm.append(value)
+            pos_inds_gmm = torch.cat(pos_inds_gmm)
+            pos_loss_gmm = torch.cat(pos_loss_gmm)
+            # fix gmm need at least two sample
+            if len(pos_inds_gmm) < 2:
+                continue
+            device = pos_inds_gmm.device
+            pos_loss_gmm, sort_inds = pos_loss_gmm.sort()
+            pos_inds_gmm = pos_inds_gmm[sort_inds]
+            pos_loss_gmm = pos_loss_gmm.view(-1, 1).cpu().numpy()
+            min_loss, max_loss = pos_loss_gmm.min(), pos_loss_gmm.max()
+            means_init = [[min_loss], [max_loss]]
+            weights_init = [0.5, 0.5]
+            precisions_init = [[[1.0]], [[1.0]]]
+            if skm is None:
+                raise ImportError('Please run "pip install sklearn" '
+                                  'to install sklearn first.')
+            gmm = skm.GaussianMixture(
+                2,
+                weights_init=weights_init,
+                means_init=means_init,
+                precisions_init=precisions_init)
+            gmm.fit(pos_loss_gmm)
+            components = gmm.predict(pos_loss_gmm)
+            scores = gmm.score_samples(pos_loss_gmm)
+            components = torch.from_numpy(components).to(device)
+            scores = torch.from_numpy(scores).to(device)
+            fgs = components == 0
+            # The implementation is (c) in Fig.3 in origin paper intead of (b).
+            # You can refer to issues such as
+            # https://github.com/kkhoot/PAA/issues/8 and
+            # https://github.com/kkhoot/PAA/issues/9.
+            if fgs.nonzero().numel():
+                _, pos_thr_ind = scores[fgs].topk(1)
+                pos_inds_after_paa.append(pos_inds_gmm[fgs][:pos_thr_ind + 1])
+
+        pos_inds_after_paa = torch.cat(pos_inds_after_paa)
+        reassign_ids = set(pos_inds.tolist()) - set(
```

Comment:
`filter_ids = (pos_inds != pos_inds_after_pss.unsqueeze(1)).all(1); reassign_ids=pos_inds[filter_ids]` should be more efficient since it's pure tensor operations and no device transfer is involved.